File: Baekjoon/yeongho/Week_2/3_OrganicCabbage.py
import sys
from collections import deque


class OrganicCabbage:
    def __init__(self):
        self.simulations = deque()

    # 입력 및 초기화
    def initialize(self):
        n = int(input())
        for _ in range(n):
            width, height, cabbage_cnt = map(int, sys.stdin.readline().split())
            field = [[0] * width for _ in range(height)]

            for _ in range(cabbage_cnt):
                x, y = map(int, sys.stdin.readline().split())
                field[y][x] = 1

            self.simulations.append(FieldSimulation(field, width, height))

    def Play(self):
        for simulation in self.simulations:
            simulation.Start()
            simulation.print_result()


class FieldSimulation:
    dx = [0, 0, -1, 1]
    dy = [1, -1, 0, 0]

    def __init__(self, field, width, height):
        self.field = field
        self.width = width
        self.height = height
        self.visited = [[False] * width for _ in range(height)]
        self.ans = 0

    def Start(self):
        for r in range(self.height):
            for c in range(self.width):
                if self.Movable(c, r):
                    self.DFS(c, r)

    def DFS(self, colum, row):
        qCabbages = deque()
        qCabbages.append((colum, row))
        self.visited[row][colum] = True

        while qCabbages:
            (x, y) = qCabbages.popleft()
            for i in range(4):
                nx, ny = (x + self.dx[i], y + self.dy[i])
                if self.InRange(nx, ny) and self.Movable(nx, ny):
                    self.visited[ny][nx] = True
                    qCabbages.append((nx, ny))

        self.ans += 1

    def InRange(self, x, y):
        return 0 <= x < self.width and 0 <= y < self.height

    def Movable(self, x, y):
        return not self.visited[y][x] and self.field[y][x] == 1

    def print_result(self):
        print(self.ans)


if __name__ == "__main__":
    tool = OrganicCabbage()

    tool.initialize()

    tool.Play()

File: Baekjoon/yeongho/Week_2/3_OrganicCabbage.cs
public class Program
{
    public static void Main()
    {
        OrganicCabbage tool = new OrganicCabbage();

        //초기화 및 입력
        tool.Initialize();
        
        // 시뮬레이션 시작
        tool.Play();
    }
}

public class OrganicCabbage
{
    private Queue<FieldSimulation> _simulations;
    
    //입력 및 초기화
    public void Initialize()
    {
        _simulations = new Queue<FieldSimulation>();
        
        int testCase = int.Parse(Console.ReadLine());
        for (int i = 0; i < testCase; i++)
        {
            var dimension = Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
            int width = dimension[0];
            int height = dimension[1];
            int cabbageCnt = dimension[2];
            int[,] field = new int[height, width];

            for (int j = 0; j < cabbageCnt; j++)
            {
                int[] pos = Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
                var (x, y) = (pos[0], pos[1]);
                field[y, x] = 1;
            }
            
            _simulations.Enqueue(new FieldSimulation(field, height, width));
        }
    }

    public void Play()
    {
        while (_simulations.Count != 0)
        {
            var simulation = _simulations.Dequeue();
            simulation.Start();
            simulation.PrintResult();
        }
    }
    
    public class FieldSimulation
    {
        private int[,] _field;
        private bool[,] _visited;
        private int RowSize;
        private int ColumSize;
        private int ans;
        // 상하좌우
        public readonly int[] dx = new[] { 0, 0, -1, 1 };
        public readonly int[] dy = new[] { 1, -1, 0, 0 };
        
        public FieldSimulation(int[,] field, int rowSize, int colSize)
        {
            _field = field;
            RowSize = rowSize;
            ColumSize = colSize;
            _visited = new bool[rowSize,colSize];
        }

        public void Start()
        {
            for (int r = 0; r < RowSize; r++)
            {
                for (int c = 0; c < ColumSize; c++)
                {
                    if(IsMovable(c,r))
                        DFS(c,r);
                }
            }
        }

        private void DFS(int c, int r)
        {
            Queue<(int, int)> qCabbages = new Queue<(int, int)>();
            qCabbages.Enqueue((c,r));
            _visited[r, c] = true;
            
            while (qCabbages.Count != 0)
            {
                var (x, y) = qCabbages.Dequeue();
                for (int i = 0; i < 4; i++)
                {
                    var nx = x + dx[i];
                    var ny = y + dy[i];

                    if (!InRange(nx, ny)) continue;
                    if (!IsMovable(nx, ny)) continue;

                    _visited[ny, nx] = true;
                    qCabbages.Enqueue((nx, ny));
                }
            }
            
            ans += 1;
        }

        private bool InRange(int x, int y)
        {
            return x >= 0 && x < ColumSize && y >= 0 && y < RowSize;
        }

        private bool IsMovable(int x, int y)
        {
            return !_visited[y, x] && _field[y,x] == 1;
        }

        public void PrintResult()
        {
            Console.WriteLine(ans);
        }
    }
    
}